OpenRA.Mods.Common/Activities/Resupply.cs

```diff
@@ -0,0 +1,210 @@
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System;
+using System.Linq;
+using OpenRA.Activities;
+using OpenRA.Mods.Common.Traits;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Activities
+{
+	public class Resupply : Activity
+	{
+		readonly IHealth health;
+		readonly RepairsUnits[] allRepairsUnits;
+		readonly Target host;
+		readonly WDist closeEnough;
+		readonly Repairable repairable;
+		readonly RepairableNear repairableNear;
+		readonly Rearmable rearmable;
+		readonly INotifyResupply[] notifyResupplies;
+
+		int remainingTicks;
+		bool played;
+		ResupplyType activeResupplyTypes = ResupplyType.None;
+
+		public Resupply(Actor self, Actor host, WDist closeEnough)
+		{
+			this.host = Target.FromActor(host);
+			this.closeEnough = closeEnough;
+			allRepairsUnits = host.TraitsImplementing<RepairsUnits>().ToArray();
+			health = self.TraitOrDefault<IHealth>();
+			repairable = self.TraitOrDefault<Repairable>();
+			repairableNear = self.TraitOrDefault<RepairableNear>();
+			rearmable = self.TraitOrDefault<Rearmable>();
+			notifyResupplies = host.TraitsImplementing<INotifyResupply>().ToArray();
+
+			var cannotRepairAtHost = health == null || health.DamageState == DamageState.Undamaged
+				|| !allRepairsUnits.Any()
+				|| ((repairable == null || !repairable.Info.RepairActors.Contains(host.Info.Name))
+					&& (repairableNear == null || !repairableNear.Info.RepairActors.Contains(host.Info.Name)));
+
+			if (!cannotRepairAtHost)
+				activeResupplyTypes |= ResupplyType.Repair;
+
+			var cannotRearmAtHost = rearmable == null || !rearmable.Info.RearmActors.Contains(host.Info.Name) || rearmable.RearmableAmmoPools.All(p => p.FullAmmo());
+			if (!cannotRearmAtHost)
+				activeResupplyTypes |= ResupplyType.Rearm;
+		}
+
+		protected override void OnFirstRun(Actor self)
+		{
+			if (host.Type == TargetType.Invalid)
+				return;
+
+			if (activeResupplyTypes > 0)
+				foreach (var notifyResupply in notifyResupplies)
+					notifyResupply.BeforeResupply(host.Actor, self, activeResupplyTypes);
+
+			// Reset the ReloadDelay to avoid any issues with early cancellation
+			// from previous reload attempts (explicit order, host building died, etc).
+			// HACK: this really shouldn't be managed from here
+			if (activeResupplyTypes.HasFlag(ResupplyType.Rearm))
+				foreach (var pool in rearmable.RearmableAmmoPools)
+					pool.RemainingTicks = pool.Info.ReloadDelay;
+		}
+
+		public override Activity Tick(Actor self)
+		{
+			if (ChildActivity != null)
+			{
+				ChildActivity = ActivityUtils.RunActivity(self, ChildActivity);
+				if (ChildActivity != null)
+					return this;
+			}
+
+			// If the activity is cancelled while we're already resupplying (or about to start resupplying),
+			// move actor outside the resupplier footprint
+			if (IsCanceling && host.IsInRange(self.CenterPosition, closeEnough))
```

Comment:
Can we add a `// HACK: This check is nowhere near robust enough, and should be rewritten`?

Reply:
Done.